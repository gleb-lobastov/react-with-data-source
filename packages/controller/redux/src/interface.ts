// eslint-disable-next-line no-unused-vars
import { READY_STATE } from './consts';

interface TRequestMeta {
  readyState: READY_STATE;
}

export interface TAction {
  type: string | symbol;
}

export interface TRequestStrategy<TRequirements extends {}, TResponse> {
  (requirements: TRequirements, ...args: any[]): Promise<TResponse>;
}

export interface TMiddlewareOptions<TRequirements extends {}, TResponse> {
  requestStrategy: TRequestStrategy<TRequirements, TResponse>;
}

export interface TDispatch {
  (action: TAction): any;
}

export interface TGetState {
  (): any;
}

export interface TStore {
  dispatch: TDispatch;
  getState: TGetState;
}

export interface TActionHandler {
  (action: TAction): any;
}

export interface TInitiateRequestAction<TRequirements extends {}>
  extends TAction {
  payload: TRequirements;
}

export interface TInvalidateRequestAction<TInvalidateRequirements>
  extends TAction {
  meta: TInvalidateRequirements;
}

export interface TRequestAction<TRequirements extends {}, TResponse>
  extends TAction {
  meta: TRequestMeta & TRequirements;
  payload?: TResponse;
  error?: boolean;
}

export interface TRequestPendingAction<TRequirements extends {}>
  extends TRequestAction<TRequirements, never> {
  meta: TRequestMeta & TRequirements;
  payload?: never;
  error?: never;
}

export interface TRequestSuccessAction<TRequirements extends {}, TResponse>
  extends TRequestAction<TRequirements, TResponse> {
  payload: TResponse;
  error?: false;
}

export interface TRequestFailureAction<TRequirements extends {}>
  extends TRequestAction<TRequirements, Error> {
  payload: Error;
  error: true;
}

export interface TRequestState<TResponse> {
  readyState: READY_STATE;
  lastError?: Error;
  lastResult?: TResponse;
  isError: boolean;
  isValid: boolean;
}
