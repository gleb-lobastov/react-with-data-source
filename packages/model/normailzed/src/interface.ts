import { IModelRequest, IModelResponsePromise } from './Model';

export interface IRequestHandler {
  (request: IModelRequest, ...args: any[]): IModelResponsePromise;
}

export interface ICustomRequest {}
