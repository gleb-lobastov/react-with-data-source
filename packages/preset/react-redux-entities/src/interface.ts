import { IModelsConfig } from '@request-kit/intg-redux-model-normalized';

export interface IReactReduxEntitiesPresetOptions {
  distributorConfig?: object;
  modelsConfig: IModelsConfig;
  requestEngineConfig?: { plugins?: any[] };
  selectorRoot?: string;
}
