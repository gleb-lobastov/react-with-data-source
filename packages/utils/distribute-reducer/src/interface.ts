export type IDomainPath = string;
export type IDomainFilterRegexp = string;

export interface IDomainFilter {
  (domainPath: IDomainPath): boolean;
}

export interface IDistributeReducerOptions {
  maxNestingDepth?: number;
  domainsFilter?: IDomainFilterRegexp;
}

export interface IActionMeta {
  domain?: IDomainPath;
  distributeReducerOptions?: IDistributeReducerOptions;
}

export interface IAction {
  meta?: IActionMeta;
}
export interface IReducer<IState> {
  (state: IState | undefined, action: IAction): IState;
}

export interface IDomainCompositeState<IDomainState> {
  domains: { [key: string]: IDomainCompositeState<IDomainState> | undefined };
  domainState?: IDomainState;
}

export interface IStepParams<IDomainState> {
  domainCompositeState: IDomainCompositeState<IDomainState> | undefined;
  domainPath: IDomainPath;
  depth: number;
}

export interface IDomainStatesMapping<IDomainState> {
  [key: string]: IDomainState;
}

export interface ISelectorOptions<IEmptyState> {
  emptyState?: IEmptyState | {};
  domainsFilter?: IDomainFilterRegexp;
}

export interface IAdaptedSelectorOptions<IEmptyState> {
  emptyState?: IEmptyState | {};
  domainsFilter?: IDomainFilter;
}
