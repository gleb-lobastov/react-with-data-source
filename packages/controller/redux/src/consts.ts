/* eslint-disable no-unused-vars */
export enum READY_STATE {
  UNSENT = 0,
  OPENED = 1,
  HEADERS_RECEIVED = 2,
  LOADING = 3,
  DONE = 4,
}
/* eslint-enable no-unused-vars */

export const EMPTY_STATE = {
  readyState: READY_STATE.UNSENT,
  isError: false,
  isValid: true,
};
