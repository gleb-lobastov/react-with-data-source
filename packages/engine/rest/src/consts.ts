export default {
  METHOD: {
    GET: 'get',
    POST: 'post',
    PATCH: 'patch',
    PUT: 'put',
    DELETE: 'delete',
    OPTIONS: 'options',
    HEAD: 'head',
  },
};
