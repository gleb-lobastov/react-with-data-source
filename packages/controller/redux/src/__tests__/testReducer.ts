import createRequestReducer from '../reducer';
import {
  createFailureAction,
  createInvalidateRequestAction,
  createPendingAction,
  createSuccessAction,
} from '../actionCreators';
import { EMPTY_STATE } from '../consts';
import { TRequestState, TAction } from '../interface'; // eslint-disable-line no-unused-vars

type TTestResponse = void;
let requestReducer: (
  state: TRequestState<TTestResponse>,
  action: TAction,
) => TRequestState<TTestResponse>;
beforeEach(() => {
  requestReducer = createRequestReducer();
});

const requirements = {};
const pendingAction = createPendingAction(requirements);
const failureAction = createFailureAction(requirements, new Error('=('));
const successAction = createSuccessAction(requirements, 'result');
const invalidateAction = createInvalidateRequestAction(requirements);
const unrelatedAction = { type: 'unrelated' };

const match = (...actions: Array<TAction>) => {
  let state = EMPTY_STATE;
  return actions.forEach(action => {
    state = requestReducer(state, action);
    expect(state).toMatchSnapshot();
  });
};
describe('request reducer', () => {
  it('should match empty state', () => {
    // @ts-ignore check for state initial request
    expect(requestReducer(undefined, unrelatedAction)).toEqual(EMPTY_STATE);
  });

  it('should match pendingAction, successAction, invalidateAction', () => {
    match(pendingAction, successAction, invalidateAction);
  });

  it('should match pendingAction, failureAction, invalidateAction', () => {
    match(pendingAction, failureAction, invalidateAction);
  });

  it('should match invalidateAction', () => {
    match(invalidateAction);
  });

  it('should match pendingAction, failureAction, pendingAction, failureAction', () => {
    match(pendingAction, failureAction, pendingAction, failureAction);
  });

  it('should match pendingAction, failureAction, pendingAction, successAction', () => {
    match(pendingAction, failureAction, pendingAction, successAction);
  });

  it('should match pendingAction, successAction, pendingAction, failureAction', () => {
    match(pendingAction, successAction, pendingAction, failureAction);
  });

  it('should match pendingAction, successAction, pendingAction, successAction', () => {
    match(pendingAction, successAction, pendingAction, successAction);
  });

  it('should match pendingAction, successAction, invalidateAction, pendingAction, successAction', () => {
    match(
      pendingAction,
      successAction,
      invalidateAction,
      pendingAction,
      successAction,
    );
  });
});
