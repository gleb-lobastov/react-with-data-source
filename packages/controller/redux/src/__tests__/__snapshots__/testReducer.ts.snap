// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request reducer should match invalidateAction 1`] = `
Object {
  "isError": false,
  "isValid": false,
  "readyState": 0,
}
`;

exports[`request reducer should match pendingAction, failureAction, invalidateAction 1`] = `
Object {
  "isError": false,
  "isValid": true,
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, failureAction, invalidateAction 2`] = `
Object {
  "isError": true,
  "isValid": true,
  "lastError": [Error: =(],
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, failureAction, invalidateAction 3`] = `
Object {
  "isError": true,
  "isValid": false,
  "lastError": [Error: =(],
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, failureAction 1`] = `
Object {
  "isError": false,
  "isValid": true,
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, failureAction 2`] = `
Object {
  "isError": true,
  "isValid": true,
  "lastError": [Error: =(],
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, failureAction 3`] = `
Object {
  "isError": true,
  "isValid": true,
  "lastError": [Error: =(],
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, failureAction 4`] = `
Object {
  "isError": true,
  "isValid": true,
  "lastError": [Error: =(],
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, successAction 1`] = `
Object {
  "isError": false,
  "isValid": true,
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, successAction 2`] = `
Object {
  "isError": true,
  "isValid": true,
  "lastError": [Error: =(],
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, successAction 3`] = `
Object {
  "isError": true,
  "isValid": true,
  "lastError": [Error: =(],
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, failureAction, pendingAction, successAction 4`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastError": [Error: =(],
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction 1`] = `
Object {
  "isError": false,
  "isValid": true,
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction 2`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction 3`] = `
Object {
  "isError": false,
  "isValid": false,
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction, pendingAction, successAction 1`] = `
Object {
  "isError": false,
  "isValid": true,
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction, pendingAction, successAction 2`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction, pendingAction, successAction 3`] = `
Object {
  "isError": false,
  "isValid": false,
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction, pendingAction, successAction 4`] = `
Object {
  "isError": false,
  "isValid": false,
  "lastResult": "result",
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, successAction, invalidateAction, pendingAction, successAction 5`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, failureAction 1`] = `
Object {
  "isError": false,
  "isValid": true,
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, failureAction 2`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, failureAction 3`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, failureAction 4`] = `
Object {
  "isError": true,
  "isValid": true,
  "lastError": [Error: =(],
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, successAction 1`] = `
Object {
  "isError": false,
  "isValid": true,
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, successAction 2`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 4,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, successAction 3`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 1,
}
`;

exports[`request reducer should match pendingAction, successAction, pendingAction, successAction 4`] = `
Object {
  "isError": false,
  "isValid": true,
  "lastResult": "result",
  "readyState": 4,
}
`;
